### PR TITLE
Fix new dashboard URLs creation

### LINKED ISSUE #2542: Incorrect new dashboard urls generation
It seems we're using `account_host` field from `app_config.yml` instead of `session_domain` to build the urls.

The account host field is only for the screens not belonging to CartoDB Editor (e.g. user options) but to Central. Session domain must be used as it represents the Editor's URL domain.

Branch is [issue-2542](https://github.com/CartoDB/cartodb/tree/issue-2542)

### PR DESCRIPTION
**Pull request by viddo**
Fixes #2542 

As per agreement with @Kartones, user and vis models will get a `base_url` attribute provided with the rest of the attributes from server-side. URL creation is now done through `.viewUrl()` on models, so far only been required for `User` and `Visualization` models so only added for those. Added a small doc page to explain this + examples, see [doc/frontend/urls.md](/cartodb/cartodb/doc/frontend/urls.md]).

From previous discussions I've also fixed:
- "Public profile" URL now points to `/maps`, due to change in URL redirects, [see #2256](https://github.com/CartoDB/cartodb/issues/2256#issuecomment-77140559)
- Set `upgrade_url` to undefined if not run as our SaaS service, [see #2646](https://github.com/CartoDB/cartodb/pull/2646)

Also, while at it, I did some cleanup related to URLs:
- Simplified router and URL creation, completely opaque to where it is actually, should be more intuitive to understand now.
- `vis.viewUrl` replaced with new url object (backward compatible).
- Fix new application layout (was failing from previous commit), extracted new private header to the layout than having it inside individual pages
- Notifications view, `.upgradeContactEmail` was moved from the urls to user model, logically makes more sense since not really a url to begin with...
- Simplified tests, removed `rewirefy` browserify plugin. One can easily mock a constructor (see changes for example) to verify params are set as expected.

### CODE REVIEW

**Comment by viddo** [comment]
@xavijam 
cc @Kartones since he brought up the issue

**Comment by viddo** [comment]
cc @Xatpy for upgrade_url check

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2138/))

File: lib/assets/javascripts/cartodb/new_common/dialogs/create/listing/imports/import_arcgis_selected_dataset_view.js [hunk 1]
```diff
@@ -23,7 +23,7 @@ module.exports = SelectedDataset.extend({
       title = title && title.replace('.' + ext, '');
     }
 
-    var upgradeAccountUrl = this.currentUserUrl.toUpgradeAccount();
+    var upgradeAccountUrl = window.upgrade_url;
     var userCanSync = this.user.get('actions') && this.user.get('actions').sync_tables;
     var customInstall = cdb.config.get('cartodb_com_hosted');
```
Comment (Xatpy) [inline lib/assets/javascripts/cartodb/new_common/dialogs/create/listing/imports/import_arcgis_selected_dataset_view.js]: Here, I'd follow your [comments](https://github.com/CartoDB/cartodb/pull/2646/files#diff-3c7d915447e7bb476de72719fc51ad12R24) in my last PR about having the same variable names (although I know that var customInstall is being used in other places)

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2139/))

File: lib/assets/javascripts/cartodb/new_common/dialogs/create/listing/imports/import_arcgis_selected_dataset_view.js (hunk 1 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb/new_common/dialogs/create/listing/imports/import_arcgis_selected_dataset_view.js]: Good catch :+1:

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2140/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2144/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2145/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2147/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2149/))

**Comment by viddo** [comment]
@juanignaciosl @rafatower since @Kartones is on vacation for unknown amount of day, can you have a review the server-side changes? Basically, [adding `base_url` property to datasets returned to client-side](https://github.com/CartoDB/cartodb/commit/13a1d39232c9a215dc3b439136ce7cbc41c517aa). Also fixed tests that were failing due to those changes. See PR desc for full scope.

**Comment by juanignaciosl** [comment]
Nothing to say here :+1:

**Comment by rafatower** [comment]
I'm ok with backend changes :+1:

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2152/))

**Comment by viddo** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2156/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2157/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2160/))

**Comment by Kartones** [comment]
fine with the BE changes :+1:

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2168/))

**Comment by viddo** [comment]
@xavijam fixed last issue found on staging, so should be good to go now. Anything more you want to test?

cc @CartoDB/frontend FYI

**Comment by xavijam** [comment]
For me :+1:

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2176/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2178/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2182/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/2184/))

**Pull request merged**